debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/OffsetContextMessageProcessingIT.java

```diff
@@ -0,0 +1,575 @@
+                        "INSERT INTO " + TEST_TABLE_SCENARIO4 + " (data, batch_id) VALUES ('scenario4-batch3-msg2', 3);" +
+                        "COMMIT;");
+
+        // Phase 3: Restart connector
+        // Should process:
+        // 1. Remaining message from batch1 (msg4) - PostgreSQL replays complete batch1, skips processed msgs
+        // 2. All messages from batch2 and batch3 (new transactions)
+        logInterceptor.clear();
+        start(PostgresConnector.class, config);
+        assertConnectorIsRunning();
+        waitForStreamingToStart();
+
+        // Should process: batch1-msg4, batch2-msg1, batch2-msg2, batch3-msg1, batch3-msg2
+        consumer = testConsumer(5);
+        consumer.await(TestHelper.waitTimeForRecords(), TimeUnit.SECONDS);
+
+        // Verify all remaining messages are processed in correct order
+        SourceRecord record4 = consumer.remove(); // batch1-msg4
+        SourceRecord record5 = consumer.remove(); // batch2-msg1
+        SourceRecord record6 = consumer.remove(); // batch2-msg2
+        SourceRecord record7 = consumer.remove(); // batch3-msg1
+        SourceRecord record8 = consumer.remove(); // batch3-msg2
+
+        // Verify batch 1 completion
+        VerifyRecord.isValidInsert(record4, PK_FIELD, 4);
+        assertThat(((Struct) record4.value()).getStruct("after").getString("data")).isEqualTo("scenario4-batch1-msg4");
+        assertThat(((Struct) record4.value()).getStruct("after").getInt32("batch_id")).isEqualTo(1);
+
+        // Verify batch 2
+        VerifyRecord.isValidInsert(record5, PK_FIELD, 5);
+        VerifyRecord.isValidInsert(record6, PK_FIELD, 6);
+        assertThat(((Struct) record5.value()).getStruct("after").getString("data")).isEqualTo("scenario4-batch2-msg1");
+        assertThat(((Struct) record6.value()).getStruct("after").getString("data")).isEqualTo("scenario4-batch2-msg2");
+        assertThat(((Struct) record5.value()).getStruct("after").getInt32("batch_id")).isEqualTo(2);
+        assertThat(((Struct) record6.value()).getStruct("after").getInt32("batch_id")).isEqualTo(2);
+
+        // Verify batch 3
+        VerifyRecord.isValidInsert(record7, PK_FIELD, 7);
+        VerifyRecord.isValidInsert(record8, PK_FIELD, 8);
+        assertThat(((Struct) record7.value()).getStruct("after").getString("data")).isEqualTo("scenario4-batch3-msg1");
+        assertThat(((Struct) record8.value()).getStruct("after").getString("data")).isEqualTo("scenario4-batch3-msg2");
+        assertThat(((Struct) record7.value()).getStruct("after").getInt32("batch_id")).isEqualTo(3);
+        assertThat(((Struct) record8.value()).getStruct("after").getInt32("batch_id")).isEqualTo(3);
+
+        LOGGER.info("Scenario 4: Intermediate offset with additional statements test completed - all messages processed correctly");
+    }
+
+    /**
+     * Additional test: Verify LSN comparison logic works correctly across different scenarios
```

Comment:
I'm not sure we need to test multiple restarts. Instead I'd add a test with overlapping transactions.